pkg/ingest/setup_test.go

```diff
@@ -3,11 +3,94 @@ package ingest
 import (
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/gstruct"
 	v1 "github.com/solo-io/build/pkg/api/v1"
 )
 
 var _ = Describe("ingest config", func() {
 
+	const relativePathToSampleConfig = "../../examples/sample-solo-project.yaml"
+	var buildId = "some-build-id"
+	const nonReleaseTagValue = ""
+	var explicitBuildEnvVars = &v1.BuildEnvVars{
+		BuildId:       buildId,
+		TaggedVersion: nonReleaseTagValue,
+	}
+	var sampleConfigIngested = v1.BuildRun{
+		Spec: &v1.BuildSpec{
+			Config: &v1.BuildConfig{
+				ReleaseContainerRegistry: &v1.ContainerRegistry{
+					Registry: &v1.ContainerRegistry_Quay{
+						Quay: &v1.QuayRegistry{
+							BaseUrl:      "quay.io",
+							Organization: "solo-io",
+						},
+					},
+				},
+				TestContainerRegistry: &v1.ContainerRegistry{
+					Registry: &v1.ContainerRegistry_Gcr{
+						Gcr: &v1.GoogleContainerRegistry{
+							BaseUrl:   "gcr.io",
+							ProjectId: "solo-public-1010",
+						},
+					},
+				},
+			},
+		},
+		Config: &v1.BuildRunConfig{
+			BuildEnvVars: &v1.BuildEnvVars{
+				BuildId:       buildId,
+				TaggedVersion: nonReleaseTagValue,
+			},
+			ComputedBuildVars: &v1.ComputedBuildVars{
+				Release:         false,
+				ImageTag:        buildId,
+				ContainerPrefix: "gcr.io/solo-public-1010",
+				Version:         buildId,
+			},
+		},
+	}
+
+	Context("InitializeBuildRun", func() {
+		It("should fallback to default filename when no file specified and env var not set", func() {
+			br, err := InitializeBuildRun("", &v1.BuildEnvVars{})
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(Equal("could not parse build spec: open solo-project.yaml: no such file or directory"))
+			Expect(br).To(Equal(v1.BuildRun{}))
+		})
+		It("should read from file when provided", func() {
+			br, err := InitializeBuildRun(relativePathToSampleConfig, explicitBuildEnvVars)
+			Expect(err).ToNot(HaveOccurred())
+
+			// aside on matching large structs
+			// Gomega has MatchAllFields, but it requires you to fill in everything, including the XXXabcd fields made by protoc
+			// An alternative is to pass IgnoreExtras with MatchFields
+			// Either way, you have to write out a lot of stuff by hand
+			// This is worthwhile in some cases because you can be explicit about how fields should match and you get
+			// great, scoped feedback on errors
```

Comment:
Agree this gives scoped errors on field differences, but at the expense of:
- test readability
- ignoring extras, which could allow future bugs